marimo/_server/api/endpoints/terminal.py

```diff
@@ -0,0 +1,92 @@
+from __future__ import annotations
+
+import asyncio
+import os
+import pty
+import select
+import signal
+import subprocess
+
+from starlette.websockets import WebSocket, WebSocketDisconnect
+
+from marimo import _loggers
+from marimo._server.api.deps import AppState
+from marimo._server.model import SessionMode
+from marimo._server.router import APIRouter
+
+LOGGER = _loggers.marimo_logger()
+
+router = APIRouter()
+
+
+async def _read_from_pty(master: int, websocket: WebSocket) -> None:
+    loop = asyncio.get_running_loop()
+    try:
+        with os.fdopen(master, "rb", buffering=0) as master_file:
+            while True:
+                try:
+                    r, _, _ = await loop.run_in_executor(
+                        None, select.select, [master_file], [], [], 0.1
+                    )
+                    if not r:
+                        await asyncio.sleep(0.1)  # Prevent busy-waiting
+                        continue
+                    data = os.read(master, 1024)
+                    if not data:
+                        break
+                    await websocket.send_text(data.decode())
+                except (asyncio.CancelledError, WebSocketDisconnect):
+                    break
+    except OSError as e:
+        if e.errno == 9:  # Bad file descriptor
+            LOGGER.debug("File descriptor closed, stopping read loop")
+            return
+        raise  # Re-raise other OSErrors
+
+
+async def _write_to_pty(master: int, websocket: WebSocket) -> None:
+    with os.fdopen(master, "wb", buffering=0) as master_file:
+        while True:
+            try:
+                data = await websocket.receive_text()
+                LOGGER.debug("Received: %s", data)
+                master_file.write(data.encode())
+                master_file.flush()
+            except (asyncio.CancelledError, WebSocketDisconnect):
+                break
+
+
+@router.websocket("/ws")
+async def websocket_endpoint(websocket: WebSocket) -> None:
+    app_state = AppState(websocket)
+    if app_state.mode != SessionMode.EDIT:
+        await websocket.close()
+        return
+
+    await websocket.accept()
+    LOGGER.debug("Terminal websocket accepted")
+
+    child_pid, fd = pty.fork()
+    if child_pid == 0:
+        default_shell = os.environ.get("SHELL", "/bin/bash")
+        subprocess.run([default_shell], shell=True)  ## noqa: ASYNC221
+        return
+
+    reader_task = asyncio.create_task(_read_from_pty(fd, websocket))
```

Comment:
Instead of asking the event loop to busy loop in `_read_from_pty`, I believe you could use `asyncio.add_reader` (similar to what we do in [`Distributor`](https://github.com/marimo-team/marimo/blob/main/marimo/_utils/distributor.py#L69-L74)) to monitor `fd`. Not sure what how `asyncio` implements that method, but it's sure to be battle tested.

If you don't do this now, can you add a TODO comment for migrating to `asyncio.add_reader`?

Reply:
adding a todo, moving to `add_reader` makes the shutdown 1 second longer